src/essence/Tools/Isochrone/IsochroneTool_Query.js

```diff
@@ -0,0 +1,153 @@
+import Map_ from "../../Basics/Map_/Map_";
+import F_ from "../../Basics/Formulae_/Formulae_";
+import L_ from "../../Basics/Layers_/Layers_";
+
+const MAX_WORKERS = 8;
+
+const QueryManager = {
+    cache: {},
+    //TODO: queue jobs here
+    //(so e.g. two isochrones will load in sequence w/o slowing each other down)
+    jobQueue: [],
+    jobInProgress: false,
+    numWorkers: 0,
+    currentJob: null,
+
+    //TODO: expand these to maintain a max cache size
+    isInCache: url => QueryManager.cache[url] !== undefined,
+    addToCache: (url, data) => QueryManager.cache[url] = data,
+
+    fillUrl: function(url, tile) {
+        //NOTE: F_.populateUrl is unsafe for polar maps
+        const pxWorldBound = Map_.map.getPixelWorldBounds(tile.z);
+        const yTileWorldBound = Math.ceil(pxWorldBound.max.y / 256) - 1;
+
+        let filledUrl = url.replace('{x}', tile.x);
+        filledUrl = filledUrl.replace('{y}', yTileWorldBound - tile.y);
+        filledUrl = filledUrl.replace('{z}', tile.z);
+        return filledUrl;
+    },
+
+    getPNG: function(url) {
+        //just promisifying png.js
+        const queryPromise = new Promise((resolve, reject) => {
+            window.PNG.load(url, resolve, reject);
+        });
+
+        //PNG.JS HAS NO NATIVE ERROR CALLBACK!
+        //Here's a terrible workaround to at least keep things moving
+        const timeoutPromise = new Promise((resolve, reject) => {
+            window.setTimeout(1000, () => reject(url + ": timed out"));
+        });
+        return Promise.race([queryPromise, timeoutPromise]);
+    },
+
+    decodePNG: function(img) {
+        if(!img) return null;
+        let rgbaArr = img.decode();
+
+        const length = rgbaArr.length / 4;
+        let heights = new Float32Array(length);
+        let p = 0;
+        for(let i = 0; i < length; i++) {
+            heights[i] = F_.RGBAto32({
+                r: rgbaArr[p],
+                g: rgbaArr[p + 1],
+                b: rgbaArr[p + 2],
+                a: rgbaArr[p + 3]
+            });
+            p += 4;
+        }
+        return heights;
+    },
+
+    getTile: async function(urlTemplate, tile) {
+        const url = this.fillUrl(urlTemplate, tile);
+        if(this.isInCache(url)) {
+            return this.cache[url];
+        } else {
+            const png = await this.getPNG(url).catch(msg => {
+                console.log(msg);
+                return null;
+            });
+            const data = this.decodePNG(png);
+            this.addToCache(url, data);
+            return data;
+        }
+    }
+}
+
+const emptyFunc = () => {};
```

Comment:
maybe leave a comment here explaining why you're reusing an empty function elsewhere